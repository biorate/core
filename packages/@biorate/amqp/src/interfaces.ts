import { IConnectorConfig, IConnector } from '@biorate/connector';
import {
  AmqpConnectionManagerOptions,
  ConnectionUrl,
  CreateChannelOpts,
} from 'amqp-connection-manager';
import { IAmqpConnectionManager } from 'amqp-connection-manager/dist/types/AmqpConnectionManager';

export type IAmqpConnection = IAmqpConnectionManager;

export interface IAmqpConfig extends IConnectorConfig {
  urls: ConnectionUrl | ConnectionUrl[];
  options?: AmqpConnectionManagerOptions;
}

export type IAmqpConnector = IConnector<IAmqpConfig, IAmqpConnection>;

export type ICreateChannelOpts = CreateChannelOpts & { name: string };
