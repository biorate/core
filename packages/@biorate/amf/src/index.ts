// Internal buffer for encoding
let tempBuffer = Buffer.allocUnsafe(1024);

// Internal buffer current offset
let tempPos = 0;

// Index for string reference encoding
let encodeCounter = 0;

// Storage for string reference encoding
let encodeCache: Record<string, number> = {};

// Storage for string reference decoding
let decodeCache: string[] = [];

/**
 * @description Resize buffer twice
 * @param {number} byte
 */
function resize(byte: number) {
  tempBuffer = Buffer.concat([tempBuffer, Buffer.allocUnsafe(byte)]);
}

/**
 * @description Write data to buffer
 * @param {*} data - Data to serialize and write
 */
function write(data: any) {
  tempPos = 0;
  encodeCounter = 0;
  encodeCache = {};
  writeValue(data);
  return tempBuffer.slice(0, tempPos);
}

/**
 * @description Check type of data and write recursive
 * @param {*} data - Data to serialize and write
 */
function writeValue(data: any) {
  if (data === undefined) return writeByte(0);
  if (data === null) return writeByte(1);
  if (data === false) return writeByte(2);
  if (data === true) return writeByte(3);

  var type = typeof data;

  if (type === 'number') {
    if (data >= 0 && data < 0x10000000 && data % 1 === 0) writeInt(data);
    else writeFloat(data);
  } else if (type === 'string') {
    writeString(data);
  } else if (Array.isArray(data)) {
    writeArray(data);
  } else if (type === 'object') {
    if (data instanceof Date) writeDate(data);
    else writeObject(data);
  } else {
    throw new Error(`${type} types not supported. Please check input data.`);
  }
}

/**
 * @description Write Date to internal buffer
 * @param {Number} value - Value to write
 */
function writeDate(value: Date) {
  writeByte(8);
  writeFloat(value.getTime(), false);
}

/**
 * @description Write byte to internal buffer
 * @param {Number} value - Value to write
 */
function writeByte(value: number) {
  if (tempBuffer.length <= tempPos) resize(1);
  tempBuffer.writeUInt8(value, tempPos);
  tempPos += 1;
}

/**
 * @description Write float64 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if float marker should not be written
 */
function writeFloat(value: number, marker = true) {
  if (marker !== false) writeByte(5);
  if (tempBuffer.length <= tempPos + 8) resize(8);
  tempBuffer.writeDoubleBE(value, tempPos);
  tempPos += 8;
}

/**
 * @description Write uInt29 to internal buffer
 * @param {Number} value - Value to write
 * @param {Boolean} [marker=true] - False if int marker should not be written
 */
function writeInt(value: number, marker = true) {
  if (marker !== false) writeByte(4);
  if (value < 0x80) {
    writeByte(value);
    return;
  }
  if (value < 0x4000) {
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  if (value < 0x200000) {
    writeByte(0x80 | ((value >> 14) & 0x7f));
    writeByte(0x80 | ((value >> 7) & 0x7f));
    writeByte(value & 0x7f);
    return;
  }
  writeByte(0x80 | ((value >> 22) & 0x7f));
  writeByte(0x80 | ((value >> 15) & 0x7f));
  writeByte(0x80 | ((value >> 8) & 0x7f));
  writeByte(value & 0xff);
}

/**
 * @description Write string to internal buffer
 * @param {String} value - String to write
 * @param {Boolean} [marker=true] - False if int marker should not be wrote
 */
function writeString(value: string, marker = true) {
  let index = encodeCache[value];

  if (marker !== false) writeByte(6);

  if (index !== undefined) {
    writeInt(index << 1, false);
    return;
  }

  index = Buffer.byteLength(value);
  writeInt((index << 1) + 1, false);

  if (index) {
    encodeCache[value] = encodeCounter;
    ++encodeCounter;
    if (tempBuffer.length <= tempPos + index) resize(index);
    tempBuffer.write(value, tempPos);
    tempPos += index;
  }
}

/**
 * @description Write array to internal buffer
 * @param {Array} value - Array to write
 */
function writeArray(value: unknown[]) {
  let len = value.length;

  writeByte(9);
  writeInt((len << 1) + 1, false);
  writeString('', false);

  for (let i = 0; i < len; ++i) {
    // Should we check for function type?
    // We write size of array before, and should add one more loop for check
    // It will be expensive
    writeValue(value[i]);
  }
}

/**
 * @description Write object to internal buffer
 * @param {Object} value - Object to write
 */
function writeObject(value: Record<string, unknown>) {
  writeByte(10);
  writeInt(11, false);
  writeString('', false);

  let keys = Object.keys(value);

  for (let i = 0, l = keys.length; i < l; ++i) {
    let name = keys[i];
    let item = value[name];

    // For consistency, we should remove this check
    // But objects frequently have methods
    if (typeof item === 'function')
      throw new Error(`Function types not supported. Please check input data.`);

    writeString(name, false);
    writeValue(item);
  }

  writeString('', false);
}

/**
 * @description Deserialize AMF3 byte buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function read<T = any>(buffer: Buffer): T {
  tempPos = 0;
  decodeCache.length = 0;

  return readValue(buffer);
}

/**
 * @description Check flag type and read unserialized value
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readValue(buffer: Buffer): any {
  switch (readByte(buffer)) {
    case 0:
      return undefined;
    case 1:
      return null;
    case 2:
      return false;
    case 3:
      return true;
    case 4:
      return readInt(buffer);
    case 5:
      return readFloat(buffer);
    case 6:
      return readString(buffer);
    case 8:
      return readDate(buffer);
    case 9:
      return readArray(buffer);
    case 10:
      return readObject(buffer);
  }
}

/**
 * @description Read byte from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readByte(buffer: Buffer) {
  tempPos += 1;
  return buffer.readUInt8(tempPos - 1);
}

/**
 * @description Read Date object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readDate(buffer: Buffer) {
  tempPos += 8;
  return new Date(buffer.readDoubleBE(tempPos - 8));
}

/**
 * @description Read float64 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readFloat(buffer: Buffer) {
  tempPos += 8;
  return buffer.readDoubleBE(tempPos - 8);
}

/**
 * @description Read int29 from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readInt(buffer: Buffer) {
  let value: number;
  let byte: number;

  byte = readByte(buffer);
  if (byte < 0x80) return byte;

  value = (byte & 0x7f) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 7;
  byte = readByte(buffer);
  if (byte < 0x80) return value | byte;

  value = (value | (byte & 0x7f)) << 8;
  byte = readByte(buffer);
  return value | byte;
}

/**
 * @description Read string from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readString(buffer: Buffer) {
  let len = readInt(buffer);

  if ((len & 1) === 0) return decodeCache[len >> 1];

  len = len >> 1;

  let string = '';

  if (len) {
    string = buffer.toString('utf8', tempPos, tempPos + len);
    decodeCache.push(string);
    tempPos += len;
  }

  return string;
}

/**
 * @description Read array from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readArray(buffer: Buffer) {
  let len = readInt(buffer);
  len = len >> 1;

  readString(buffer);

  let result = [];

  for (let i = 0; i < len; ++i) result.push(readValue(buffer));

  return result;
}

/**
 * @description Read object from buffer
 * @param {Buffer} buffer - AMF3 bytes buffer
 */
function readObject(buffer: Buffer) {
  let name;
  readInt(buffer);
  readString(buffer);

  let result: Record<string, unknown> = {};

  while ((name = readString(buffer))) result[name] = readValue(buffer);

  return result;
}

/**
 * @description AMF3 encode function
 *
 * Links:
 * - [wikipedia](https://en.wikipedia.org/wiki/Action_Message_Format)
 * - [specification] https://www.adobe.com/content/dam/acom/en/devnet/pdf/amf-file-format-spec.pdf
 *
 * Limitations:
 * - Function is not supported
 * - XML is not supported
 * - ByteArray is not supported
 * - Dictionary is not supported
 * - Vectors is not supported
 *
 * @example
 * ```ts
 * import { encode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(buffer); // <Buffer 0a 0b 01 09 74 65 73 74 04 01 01>
 * ```
 */
export const encode = write;

/**
 * @description AMF3 decode function
 *
 * Limitations:
 *   - see encode description section
 *
 * @example
 * ```ts
 * import { encode, decode } from '@biorate/amf';
 *
 * const buffer = encode({ test: 1 });
 *
 * console.log(decode(buffer)); // { test: 1 }
 * ```
 */
export const decode = read;
