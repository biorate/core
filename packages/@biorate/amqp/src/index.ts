import { EventEmitter } from 'events';
import { injectable } from '@biorate/inversion';
import { Connector } from '@biorate/connector';
import { events } from '@biorate/tools';
import { connect, ChannelWrapper } from 'amqp-connection-manager';
import { AmqpCantConnectError, ChannelNotExistsError } from './errors';
import { IAmqpConfig, IAmqpConnection, ICreateChannelOpts } from './interfaces';

export { ConsumeMessage } from 'amqplib/properties';
export * from 'amqp-connection-manager';
export * from './errors';
export * from './interfaces';

/**
 * @description Amqp connector
 *
 * ### Features:
 * - connection manager for AMQP
 *
 * @example
 * ```
 * import { inject, container, Types, Core } from '@biorate/inversion';
 * import { IConfig, Config } from '@biorate/config';
 * import { AmqpConnector } from '@biorate/amqp';
 *
 * const connectionName = 'amqp';
 * const channelName = 'test';
 *
 * export class Root extends Core() {
 *   @inject(AmqpConnector) public connector: AmqpConnector;
 * }
 *
 * container.bind<IConfig>(Types.Config).to(Config).inSingletonScope();
 * container.bind<AmqpConnector>(AmqpConnector).toSelf().inSingletonScope();
 * container.bind<Root>(Root).toSelf().inSingletonScope();
 *
 * container.get<IConfig>(Types.Config).merge({
 *   Amqp: [
 *     {
 *       name: connectionName,
 *       urls: ['amqp://localhost:5672'],
 *     },
 *   ],
 * });
 *
 * const root = <Root>container.get<Root>(Root);
 *
 * (async () => {
 *   root.connector.createChannel(connectionName, {
 *     name: channelName,
 *     json: true,
 *     setup: async (channel: Channel) => {
 *       await channel.assertExchange('test-exchange', 'topic');
 *       await channel.assertQueue('test-queue', { exclusive: true, autoDelete: true });
 *       await channel.bindQueue('test-queue', 'test-exchange', '#send');
 *       await channel.consume('test-queue', (data: ConsumeMessage | null) => {
 *         console.log(data?.content?.toString?.()); // {"test": 1}
 *       });
 *     },
 *   });
 *   root.connector.channel(channelName)!.publish('test-exchange', '#send', { test: 1 });
 * })();
 * ```
 */
@injectable()
export class AmqpConnector extends Connector<IAmqpConfig, IAmqpConnection> {
  /**
   * @description Private connections storage
   */
  private '#connections': Map<string, IAmqpConnection>;
  /**
   * @description Private link to selected (used) connection
   */
  private '#current': IAmqpConnection | undefined;
  /**
   * @description Namespace path for fetching configuration
   */
  protected readonly namespace = 'Amqp';
  /**
   * @description Channels storage
   */
  protected readonly channels = new Map<string, ChannelWrapper>();
  /**
   * @description Create connection
   */
  protected async connect(config: IAmqpConfig) {
    let connection: IAmqpConnection;
    try {
      connection = connect(config.urls, config.options);
      await events.once(<EventEmitter>(<unknown>connection), 'connect');
      connection.on('disconnect', console.warn);
    } catch (e: unknown) {
      throw new AmqpCantConnectError(<Error>e);
    }
    return connection;
  }
  /**
   * @description Create channel method
   */
  public createChannel(name: string, options: ICreateChannelOpts) {
    const channel = this.connection(name).createChannel(options);
    this.channels.set(options.name, channel);
    return channel;
  }
  /**
   * @description Get channel method
   */
  public channel(name: string) {
    if (!this.channels.has(name)) throw new ChannelNotExistsError(name);
    return this.channels.get(name);
  }
}
