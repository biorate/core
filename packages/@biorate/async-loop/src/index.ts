import { timer } from '@biorate/tools';

/**
 * @description Async loop implementation
 *
 * @example
 * ```
 * import { AsyncLoop } from '@biorate/async-loop';
 *
 * (async () => {
 *   let i = 0;
 *   const loop = new AsyncLoop(
 *     () => {
 *       console.log(++i); // 1, 2, 3 ... e.t.c
 *     },
 *     console.error,
 *     1000,
 *   );
 * })();
 * ```
 */
export class AsyncLoop {
  #process: () => Promise<void> | void;

  #error: (e: unknown) => Promise<void> | void;

  #interval: number;

  #paused = false;

  #stoped = false;

  public constructor(
    process: () => Promise<void> | void,
    error?: () => Promise<void> | void,
    interval = 1000,
  ) {
    this.#process = process;
    this.#error = error ?? console.error;
    this.#interval = interval;
    this.process().catch(this.#error);
  }

  public pause() {
    this.#paused = true;
  }

  public resume() {
    this.#paused = false;
  }

  public stop() {
    this.#stoped = true;
  }

  protected async process() {
    while (true) {
      await timer.wait(this.#interval);
      if (this.#paused) continue;
      if (this.#stoped) break;
      try {
        await this.#process();
      } catch (e: unknown) {
        await this.#error(e);
      }
    }
  }
}
