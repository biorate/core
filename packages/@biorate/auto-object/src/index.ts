import 'reflect-metadata';
import { flattenDeep } from 'lodash';
import { auto, transaction } from './utils';
import { PropertiesOnly } from './interfaces';

export * from './decorators';
export { Getter, Setter, PropertiesOnly } from './interfaces';
/**
 * @description
 * Auto initialized object
 *
 * @example
 * ```ts
 * class Geo extends Array<number> {}
 *
 * class Address extends AutoObject<Address> {
 *   @IsString()
 *   public city: string;
 *
 *   @IsString()
 *   public street: string;
 *
 *   @IsNumber()
 *   public apartment: number;
 *
 *   @IsNumber()
 *   public postal: number;
 *
 *   @IsArray()
 *   public geo: Geo;
 *
 *   public get inline() {
 *     return <Getter<string>>(
 *       `${this.postal}, ${this.city}, ${this.street} â„–${this.apartment}`
 *     );
 *   }
 * }
 *
 * class User extends AutoObject<User> {
 *   @IsInt()
 *   public id: number;
 *
 *   @IsString()
 *   public firstName: string;
 *
 *   @IsString()
 *   public lastName: string;
 *
 *   @IsObject()
 *   @ValueObject(Address)
 *   public address: Address;
 * }
 *
 * const user = new User({
 *   id: 1,
 *   firstName: 'Vasya',
 *   lastName: 'Pupkin',
 *   address: {
 *     city: 'Moscow',
 *     street: 'Gogolya',
 *     postal: 123321,
 *     apartment: 74,
 *     geo: [12321, 32123],
 *   },
 * });
 *
 * console.log(user); // User {
 * //   "address": Address {
 * //     "city": "Moscow",
 * //     "apartment": 74,
 * //     "geo": Array [
 * //       12321,
 * //       32123,
 * //     ],
 * //     "postal": 123321,
 * //     "street": "Gogolya",
 * //   },
 * //   "firstName": "Vasya",
 * //   "id": 1,
 * //   "lastName": "Pupkin",
 * // }
 * ```
 */
export abstract class AutoObject<T = Record<string, any>> {
  public static extends<C extends new (...args: any[]) => any, T = Record<string, any>>(
    Class: C,
  ) {
    return class AutoObject extends (Class ? Class : <C>class {}) {
      public constructor(...args: any[]) {
        super(...args.slice(1));
        return auto<PropertiesOnly<T>>(this, args[0]);
      }
    };
  }

  public constructor(data: PropertiesOnly<T>) {
    return auto<PropertiesOnly<T>>(this, data);
  }
}

export abstract class AutoArray<T> extends Array<PropertiesOnly<T>> {
  #transform = (data: PropertiesOnly<T>) => new this.Class(data);

  protected abstract get Class(): new (...args: any[]) => any;

  public constructor(...args: PropertiesOnly<T>[]) {
    super();
    this.push(...args.map(this.#transform));
  }

  public push(...args: PropertiesOnly<T>[]) {
    return super.push(...args.map(this.#transform));
  }

  public unshift(...args: PropertiesOnly<T>[]) {
    return super.unshift(...args.map(this.#transform));
  }

  public slice(start: number, deleteCount?: number): this {
    return <typeof this>transaction(() => super.slice(start, deleteCount));
  }

  public splice(start: number, deleteCount: number, ...args: PropertiesOnly<T>[]): this {
    return <typeof this>(
      transaction(() => super.splice(start, deleteCount, ...args.map(this.#transform)))
    );
  }

  public concat(...args: (PropertiesOnly<T> | ConcatArray<PropertiesOnly<T>>)[]): this {
    return <typeof this>(
      transaction(() =>
        super.concat((<PropertiesOnly<T>[]>flattenDeep(<any>args)).map(this.#transform)),
      )
    );
  }
}
