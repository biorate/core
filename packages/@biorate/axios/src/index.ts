import { omit, pick, merge } from 'lodash';
import axios, { AxiosResponse, AxiosInstance } from 'axios';
import retry from 'axios-retry';
import { IAxiosRetryConfig } from 'axios-retry/dist/esm';
// @ts-ignore
import * as pathToUrl from 'path-to-url';
import { Singleton } from '@biorate/singleton';
import { Stubs } from './stubs';
import { IAxiosFetchOptions, IStubParam } from './interfaces';

export * from 'axios';
export * from './interfaces';

const axiosRetryConfigKeys = [
  'retries',
  'retryDelay',
  'shouldResetTimeout',
  'retryCondition',
  'onRetry',
];

const axiosExcludeKeys = ['path', 'config', 'retry'];

/**
 * @description
 * Axios OOP static interface
 *
 * ### Features:
 * - OOP
 * - DI
 *
 * @example
 * ```
 * import { Axios } from '@biorate/axios';
 *
 * class Yandex extends Axios {
 *   public baseURL = 'https://yandex.ru';
 * }
 *
 * (async () => {
 *   const response = await Yandex.fetch<string>();
 *   console.log(response.status); // 200
 *   console.log(response.data); // <!DOCTYPE html><html ...
 * })();
 * ```
 */
export class Axios extends Singleton {
  /**
   * @description Axios instance cache
   */
  protected static mocks = new WeakMap<typeof Axios, { value: boolean }>();
  /**
   * @description Stubs instance cache
   */
  protected static get stubs() {
    return Stubs.get(this);
  }
  /**
   * @description Stubs options cache
   */
  public static get options() {
    return Stubs.get(this).options;
  }
  /**
   * @description Fetch static method
   */
  public static fetch(options?: any) {
    return this._fetch(options);
  }
  /**
   * @description Stub static method
   */
  public static stub(params: IStubParam, persist = false) {
    this.stubs.stub(this.instance<Axios>(), params, persist);
  }
  /**
   * @description Unstub static method
   */
  public static unstub() {
    this.stubs.unstub();
  }
  /**
   * @description Use mock static method
   */
  public static useMock() {
    this.mocks.set(this, { value: true });
  }
  /**
   * @description Set defaults
   */
  public static get defaults() {
    return axios.defaults;
  }
  /**
   * @description Get mock static method
   */
  protected static getMock<T = any, D = any>(
    instance: Axios,
    options?: IAxiosFetchOptions,
  ): undefined | AxiosResponse<T, D> {
    return undefined;
  }
  /**
   * @description Set mock static method
   */
  protected static setMock<T = any, D = any>(
    instance: Axios,
    result: AxiosResponse<T, D>,
    options?: IAxiosFetchOptions,
  ) {}
  /**
   * @description Protected fetch static method
   */
  protected static async _fetch<T = any, D = any>(
    options?: IAxiosFetchOptions,
  ): Promise<AxiosResponse<T, D>> {
    const instance = this.instance<Axios>();
    const useMock = this.mocks.get(this)?.value;
    if (useMock) {
      const mock = this.getMock<T, D>(instance, options);
      if (mock) return mock;
    }
    const result = await instance.fetch<T, D>(options);
    if (useMock) this.setMock<T, D>(instance, result, options);
    return result;
  }
  /**
   * @description Axios client cache
   */
  #client: AxiosInstance;
  /**
   * @description Default timeout limit
   */
  public timeout = 15000;
  /**
   * @description Fetch method
   */
  protected async fetch<T, D>(
    options?: IAxiosFetchOptions,
  ): Promise<AxiosResponse<T, D>> {
    const settings = merge({ ...this }, options);
    if (!this.#client) {
      this.#client = axios.create();
      if (settings.retry)
        // @ts-ignore: TODO: Stranger Things
        retry(this.#client, <IAxiosRetryConfig>pick(this, axiosRetryConfigKeys));
    }
    if (settings.baseURL && settings.path)
      settings.baseURL = pathToUrl(settings.baseURL, settings.path);
    if (settings.url && settings.path)
      settings.url = pathToUrl(settings.url, settings.path);
    const params = {
      ...omit(settings, axiosRetryConfigKeys.concat(axiosExcludeKeys)),
    };
    await this.before(params);
    const startTime = this.getStartTime();
    try {
      const result = await this.#client(params);
      await this.after<T>(result, startTime, params);
      return result;
    } catch (e: unknown) {
      await this.catch(<Error>e, startTime, params);
      throw e;
    } finally {
      await this.finally(startTime, params);
    }
  }
  /**
   * @description Get start time method
   */
  protected getStartTime(): [number, number] {
    return [Date.now(), 0];
  }
  /**
   * @description Before hook
   */
  protected async before(params?: IAxiosFetchOptions) {}
  /**
   * @description After hook
   */
  protected async after<T = any, D = any>(
    result: AxiosResponse<T, D>,
    startTime: [number, number],
    params?: IAxiosFetchOptions,
  ) {}
  /**
   * @description Catch hook
   */
  protected async catch(
    e: Error,
    startTime: [number, number],
    params?: IAxiosFetchOptions,
  ) {}
  /**
   * @description Finally hook
   */
  protected async finally(startTime: [number, number], params?: IAxiosFetchOptions) {}
}
