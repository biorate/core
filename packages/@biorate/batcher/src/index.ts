import { injectable } from '@biorate/inversion';
import { timer } from '@biorate/tools';
import { ITask, IMetadata, IBatcher } from './interfaces';

export * from './interfaces';

/**
 * @description
 * Tasks batcher
 *
 * ### Features:
 * - Group single tasks into batch request
 *
 * @example
 * ```
 * import { inject, container, Types, Core } from '@biorate/inversion';
 * import { IConfig, Config } from '@biorate/config';
 * import { IBatcher, Batcher } from '@biorate/batcher';
 *
 * const batcher: IBatcher = new Batcher<{ data: string }, { test: string }>();
 *
 * batcher.register((tasks) => {
 *   console.log(tasks);
 *   // [
 *   //   [
 *   //     { data: 'one' },
 *   //     {
 *   //       resolve: [Function (anonymous)],
 *   //       reject: [Function (anonymous)],
 *   //       metadata: { test: 'one' }
 *   //     }
 *   //   ],
 *   //   [
 *   //     { data: 'two' },
 *   //     {
 *   //       resolve: [Function (anonymous)],
 *   //       reject: [Function (anonymous)],
 *   //       metadata: { test: 'two' }
 *   //     }
 *   //   ],
 *   //   [
 *   //     { data: 'three' },
 *   //     {
 *   //       resolve: [Function (anonymous)],
 *   //       reject: [Function (anonymous)],
 *   //       metadata: { test: 'three' }
 *   //     }
 *   //   ]
 *   // ]
 * });
 * batcher.add({ data: 'one' }, { test: 'one' });
 * batcher.add({ data: 'two' }, { test: 'two' });
 * batcher.add({ data: 'three' }, { test: 'three' });
 * ```
 */
@injectable()
export class Batcher<O = unknown, M = IMetadata> implements IBatcher<O, M> {
  protected stamp = Date.now();
  protected tasks: [O, ITask<M>][] = [];
  protected callback: (tasks: [O, ITask<M>][]) => void | Promise<void>;
  protected unique = new Set<symbol | string>();

  public constructor(protected count = 100, protected timeout = 100) {
    this.loop().catch(this.panic);
  }

  public get length() {
    return this.tasks.length;
  }

  public register(callback: (tasks: [O, ITask<M>][]) => void | Promise<void>) {
    this.callback = callback;
  }

  public rollback(tasks: [O, ITask<M>][]) {
    this.tasks.unshift(...tasks);
  }

  public add(object: O, metadata?: M) {
    return new Promise<unknown>((resolve, reject) =>
      this.tasks.push([
        object,
        {
          resolve,
          reject,
          metadata,
        },
      ]),
    );
  }

  public async force() {
    const tasks = this.tasks.slice();
    this.tasks.length = 0;
    await this.callback(this.deduplicate(tasks));
  }

  protected panic(e: Error) {
    throw e;
  }

  protected async loop() {
    while (true) {
      await timer.wait();
      await this.process();
    }
  }

  protected key(object: O): symbol | string {
    return Symbol();
  }

  protected deduplicate(tasks: [O, ITask<M>][]) {
    this.unique.clear();
    for (let i = tasks.length; i--; ) {
      const key = this.key(tasks[i][0]);
      if (!this.unique.has(key)) this.unique.add(key);
      else this.tasks.push(...tasks.splice(i, 1));
    }
    return tasks;
  }

  protected async process() {
    if (!this.callback) return;
    if (this.stamp > Date.now() && this.tasks.length < this.count) return;
    this.stamp = Date.now() + this.timeout;
    if (this.tasks.length) {
      const tasks = this.tasks.slice();
      this.tasks.length = 0;
      await this.callback(this.deduplicate(tasks));
    }
  }
}
