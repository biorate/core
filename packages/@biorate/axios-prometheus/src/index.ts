import { container, Types } from '@biorate/inversion';
import { IConfig } from '@biorate/config';
import { AxiosError, AxiosResponse } from 'axios';
import { Axios } from '@biorate/axios';
import { counter, Counter, histogram, Histogram } from '@biorate/prometheus';

/**
 * @description
 * Axios-prometheus HTTP interface
 *
 * ### Features:
 * - Metrics of you HTTP requests out of the box
 *
 * @example
 * ```
 * import { container, Types } from '@biorate/inversion';
 * import { IConfig, Config } from '@biorate/config';
 * import { AxiosPrometheus } from '@biorate/axios-prometheus';
 *
 * export class Google extends AxiosPrometheus {
 *   public baseURL = this.config.get<string>('baseURL');
 *   public url = '/';
 *   public method = 'get';
 *   public timeout = 1500;
 * }
 *
 * container.bind(Types.Config).to(Config).inSingletonScope();
 *
 * container.get<IConfig>(Types.Config).merge({
 *   baseURL: 'https://google.com',
 * });
 *
 * (async () => {
 *   await Google.fetch();
 * })();
 * ```
 */
export abstract class AxiosPrometheus extends Axios {
  protected get config() {
    return container.get<IConfig>(Types.Config);
  }

  @counter({
    name: 'api_request_counter',
    help: 'api request counter',
    labelNames: ['method', 'baseUrl', 'statusCode'],
  })
  protected counter: Counter;

  @histogram({
    name: 'api_response_time',
    help: 'api response time',
    labelNames: ['method', 'baseUrl', 'statusCode'],
    buckets: [5, 10, 20, 50, 100, 300, 500, 1000, 2000, 3000, 5000, 10000],
  })
  protected histogram: Histogram;

  public abstract baseURL: string;
  public abstract url: string;
  public abstract method: string;

  protected log(statusCode: number, startTime: [number, number]) {
    const diff = process.hrtime(startTime);
    const time = diff[0] * 1e3 + diff[1] * 1e-6;
    this.counter
      .labels({
        method: this.method,
        baseUrl: this.baseURL + this.url,
        statusCode,
      })
      .inc();
    this.histogram
      .labels({
        method: this.method,
        baseUrl: this.baseURL + this.url,
        statusCode,
      })
      .observe(time);
  }

  protected async after(result: AxiosResponse, startTime: [number, number]) {
    this.log(result.status, startTime);
  }

  protected async catch(e: Error | AxiosError, startTime: [number, number]) {
    if ('response' in e) this.log(e!.response!.status, startTime);
  }
}
