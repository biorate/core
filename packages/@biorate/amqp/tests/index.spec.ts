import { expect } from 'chai';
import { Channel, ConsumeMessage } from '../src';
import { root, channelName, connectionName } from './__mocks__';

describe('@biorate/amqp', function () {
  this.timeout(3e4);

  before(async () => await root.$run());

  it('publish / consume', (done) => {
    root.connector.createChannel(connectionName, {
      name: channelName,
      json: true,
      setup: async (channel: Channel) => {
        await channel.assertExchange('test-exchange', 'topic');
        await channel.assertQueue('test-queue', { exclusive: true, autoDelete: true });
        await channel.bindQueue('test-queue', 'test-exchange', '#send');
        await channel.consume('test-queue', (data: ConsumeMessage | null) => {
          expect(data?.content?.toString?.()).toMatchSnapshot();
          done();
        });
      },
    });
    root.connector.channel(channelName)!.publish('test-exchange', '#send', { test: 1 });
  });
});
