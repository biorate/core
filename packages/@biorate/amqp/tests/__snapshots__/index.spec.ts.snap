// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@biorate/amqp publish / consume 1`] = `"{\\"test\\":1}"`;
